osquery/tables/system/windows/prefetch.cpp

```diff
@@ -0,0 +1,358 @@
+}
+
+void parsePrefetchVersion(QueryData& results,
+                          const std::string& prefetch_data,
+                          const std::string& file_path) {
+  std::string str_version = prefetch_data.substr(0, 8);
+  str_version = swapEndianess(str_version);
+  int version = tryTo<int>(str_version, 16).takeOr(0);
+  // Currently supports Win7 and higher
+  if (version == 30 || version == 23 || version == 26) {
+    parsePrefetchData(results, prefetch_data, file_path, version);
+  } else {
+    LOG(WARNING) << "Unsupported prefetch file: " << file_path;
+  }
+}
+
+void parsePrefetch(const std::vector<std::string>& prefetch_files,
+                   QueryData& results) {
+  for (const auto& file : prefetch_files) {
+    if (boost::algorithm::iends_with(file, ".pf") &&
+        boost::filesystem::is_regular_file(file)) {
+      LOG(INFO) << "Parsing prefetch file: " << file;
+      std::ifstream input_file(file, std::ios::in | std::ios::binary);
+      std::vector<char> compressed_data(
+          (std::istreambuf_iterator<char>(input_file)),
+          (std::istreambuf_iterator<char>()));
+      input_file.close();
+      int sig_size = 0;
+      std::string header_sig = "";
+      std::stringstream decom_ss;
+      // Get header information
+      for (const auto& data : compressed_data) {
+        if (sig_size == 8) {
+          break;
+        }
+        if (sig_size < 3) {
+          header_sig += data;
+        } else if (sig_size > 3) {
+          std::stringstream value;
+          value << std::setfill('0') << std::setw(2);
+          value << std::hex << std::uppercase
+                << static_cast<int>(static_cast<unsigned char>((data)));
+          decom_ss << value.str();
+        }
+        sig_size++;
+      }
+      std::string data = "";
+      std::string file_path = file;
+      // Check for compression signature. Prefetch may be compressed on Win8+
+      if (header_sig == "MAM") {
+        std::string prefetch_size = swapEndianess(decom_ss.str());
+        unsigned long size =
+            tryTo<unsigned long>(prefetch_size, 16).takeOr(0ul);
+        if (size == 0l) {
+          LOG(WARNING) << "Could not get prefetch data size for: " << file_path;
+          continue;
+        }
+        auto expected_data = decompressLZxpress(compressed_data, size);
+        if (expected_data.isError()) {
+          continue;
+        }
+        data = expected_data.take();
+      } else {
+        std::stringstream prefetch_ss;
+
+        for (const auto& compressed : compressed_data) {
+          std::stringstream value;
+          value << std::setfill('0') << std::setw(2);
+          value << std::hex << std::uppercase
+                << static_cast<int>(static_cast<unsigned char>((compressed)));
+          prefetch_ss << value.str();
+        }
+        data = prefetch_ss.str();
+      }
+      const std::string sig_header = data.substr(8, 8);
+      // Check for "SCCA" signature
+      if (sig_header != "53434341") {
```

Comment:
Isn't converting all the data to hex going to make things slow? It appears that `decompressLZxpress` is doing this.

Reply:
I based this off the shortcut_files which is doing a similar thing.  Several of the helper functions that the prefetch parsing uses also expect a hex string (swapEndianess, littleEndianToUnixTime).
Also during my testing it does not appear to be slow as mentioned below? (sorry not the best answer :sweat_smile:)

Reply:
Let me rephrase my question. Is converting the file into a hex string for processing necessary?

Reply:
yes

Reply:
Hmmm, but here in this line, couldn't the check just as likely have been `unsigned long sig_header = (unsigned long*)(&compressed_data[4]); if (sig_header == 'ACCS') ...` in bytes vs converting the contents of the entire file to a hex string? With respect to `littleEndianToUnixTime`, there is also a `longIntToUnixtime` that does not require endian conversion, for example.

Reply:
Sorry, don't want to pile on, but checking bytes vs a string, especially when this string is not something human readable anyway, seems better, especially to reduce the amount of allocations needed.

I wanted also to suggest, in the case strings remain, to have a look at `std::string_view` (C++17), to avoid having to make continuous copies/substr of the original data, especially if it's information that it's not stored into a row.

Reply:
So i took a look at `string_view` and encountered some issues?
It looks like once `string_view` is used its no longer possible to use strings?
Several helper functions this table uses expect a string `tryTo, swapEndianess, littleEndianToUnixTime`
One of the reasons im using strings is because prefetch files can contain remnant/extra data. Using strings makes it a bit easier to find the correct offsets for the data.

All this feedback is really great and i want this table to be implemented the right way. So if bytes are strongly preferred over strings just let me know and i try to rewrite the parsing to use bytes.

Reply:
This should be resolved, the table now parses the binary data and not strings.  The parsing should be much faster